pandas/tests/groupby/test_apply.py

```diff
@@ -101,10 +101,43 @@ def f(g):
     splitter = grouper._get_splitter(g._selected_obj, axis=g.axis)
     group_keys = grouper._get_group_keys()
 
-    values, mutated = splitter.fast_apply(f, group_keys)
+    values, mutated, succsessful_apply = splitter.fast_apply(f, group_keys)
+    # The bool successful_apply signals whether or not the fast apply was
+    # successful on the entire data set. It is false for cases which need to
+    # fall back to a slow apply code path for safety reasons.
+    assert succsessful_apply
     assert not mutated
 
 
+def test_group_apply_once_per_group():
+    # GH2936, GH7739, GH10519, GH2656, GH12155, GH20084, GH21417
+
+    # This test should ensure that a function is only evaluted
+    # once per group. Previously the function has been evaluated twice
```

Comment:
this really covers *all* of the above issues? meaning they all have exactly the same input? I find that hard to believe. Please parameterize as needed, but run thru the cases.

Reply:
The test cases are in fact pretty close to each others. To make the review easier I included all test cases but I believe most are redundant. If you see no need for the duplication, I can easily remove / shrink the test again